packages/next/src/lib/metadata/resolve-metadata.ts

```diff
@@ -248,40 +230,79 @@ export async function collectMetadata(
+  let resolvedTwitterTitle: AbsoluteTemplateString | null = null
+  let resolvedOpenGraphTitle: AbsoluteTemplateString | null = null
+
+  // Loop over all metadata items again, merging synchronously any static object exports,
+  // awaiting any static promise exports, and resolving parent metadata and awaiting any generated metadata
 
-  for (const item of metadataItems) {
-    const [metadataExport, staticFilesMetadata] = item
-    const currentMetadata =
-      typeof metadataExport === 'function'
-        ? metadataExport(parentPromise)
-        : metadataExport
-    const layerMetadataPromise =
-      currentMetadata instanceof Promise
-        ? currentMetadata
-        : Promise.resolve(currentMetadata)
+  let resolvingIndex = 0
+  for (let i = 0; i < metadataItems.length; i++) {
+    const [metadataExport, staticFilesMetadata] = metadataItems[i]
+    let metadata: Metadata | null = null
+    if (typeof metadataExport === 'function') {
+      // call each `generateMetadata function concurrently and stash their resolver
+      generateMetadataResults.push(
+        metadataExport(
+          new Promise((resolve) => {
+            resolvers.push(resolve)
+          })
+        )
+      )
 
-    parentPromise = parentPromise.then((resolved) => {
-      return layerMetadataPromise.then((metadata) => {
-        resolved = deepClone(resolved)
-        merge(resolved, metadata, staticFilesMetadata, {
-          title: resolved.title?.template || null,
-          openGraph: resolved.openGraph?.title?.template || null,
-          twitter: resolved.twitter?.title?.template || null,
-        })
-        return process.env.NODE_ENV === 'development'
-          ? Object.freeze(resolved)
-          : resolved
-      })
-    })
+      const resolveParent = resolvers[resolvingIndex]
+      const generatedMetadata = generateMetadataResults[resolvingIndex]
+      resolvingIndex++
+      // In dev we clone and freeze to prevent relying on mutating resolvedMetadata directly.
+      // In prod we just pass resolvedMetadata through without any copying.
+      const currentResolvedMetadata: ResolvedMetadata =
+        process.env.NODE_ENV === 'development'
+          ? Object.freeze(
+              require('next/dist/compiled/@edge-runtime/primitives/structured-clone').structuredClone(
+                resolvedMetadata
+              )
+            )
+          : resolvedMetadata
+
+      // This resolve should unblock the generateMetadata function if it awaited the parent
+      // argument. If it didn't await the parent argument it might already have a value since it was
+      // called concurrently. Regardless we await the return value before continuing on to the next layer
+      resolveParent(currentResolvedMetadata)
+      metadata =
+        generatedMetadata instanceof Promise
+          ? await generatedMetadata
+          : generatedMetadata
+    } else {
+      metadata = metadataExport
+    }
```

Comment:
```suggestion
    } else if (metadataExport !== null && typeof metadataExport === 'object') {
      // This metadataExport is the object form
      metadata = metadataExport
    } else {
      // there is no metadata export at this layout/page, short circuit the loop and continue
      continue;
    }
```

Reply:
can't do continue since there might be some merging for static metadata. I'll keep the `else if` branch in incoming updates